topics/go/algorithms/data/hash/hash.go

```diff
@@ -0,0 +1,107 @@
+// Package hash implements a hash table
+package hash
+
+import (
+	"fmt"
+	"hash/maphash"
+)
+
+const (
+	numBuckets = 256 // Number of buckets in hash
+)
+
+// entry in the hash table
+type entry struct {
+	key   string
+	value int
+}
+
+// Hash is a simple Hash table
+type Hash struct {
+	buckets [][]*entry
+	h       maphash.Hash
+}
+
+// NewHash returns a new hash table
+func NewHash() *Hash {
+	return &Hash{
+		buckets: make([][]*entry, numBuckets),
```

Comment:
Why do we need to use pointer semantics in the collection. I am going back and forth on this. If the hash table is holding the original I would want to use value but if this is a secondary index, then pointer would be good.  Not Sure.

Reply:
The pointer semantics is for `Set`. It's either what's currently in the code have or:
```go
	b := h.hashKey(key)

	// Look for existing key
	for i := range h.buckets[b] {
		if h.buckets[i].key == key {
			(&h.buckets[i]).value = value
			return
		}
	}

	h.buckets[b] = append(h.buckets[b], &entry{key, value})
```

The line `(&h.buckets[i]).value = value` is tricky. Using the following won't work.
```go
for _, e := range h.buckets[b] {
		if e.key == key {
			(&e).value = value
			return
```

Your call, I'm good with both.

Reply:
Can we use value semantic mutation and just replace the value?

Reply:
I feel like I am missing something. We could just copy over the value for the new one?